### PR TITLE
KAFKA-7778: Add KTable.suppress to Scala API

### PR DESCRIPTION
**Pull request by ouertani**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*
Add KTable.suppress to Scala API
https://issues.apache.org/jira/browse/KAFKA-7778

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [X] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [X ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
Jenkins failure seems related to `spotless`. Could you run unit test locally and see if it passed? Also could you add the corresponding unit test?

**Comment by guozhangwang** [comment]
cc @vvcephei

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala [hunk 1]
```diff
@@ -159,6 +159,16 @@ class KTable[K, V](val inner: KTableJ[K, V]) {
   def toStream[KR](mapper: (K, V) => KR): KStream[KR, V] =
     inner.toStream[KR](mapper.asKeyValueMapper)
 
+  /**
+    * Suppress some updates from this changelog stream, determined by the supplied {@link Suppressed} configuration.
```
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala]: `@link` is Javadoc, not valid Scaladoc. In scaladoc, it's `[[Suppressed]]`:

```suggestion
    * Suppress some updates from this changelog stream, determined by the supplied [[Suppressed]] configuration.
```

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala [hunk 2]
```diff
@@ -159,6 +159,16 @@ class KTable[K, V](val inner: KTableJ[K, V]) {
...
+    *
+    * This controls what updates downstream table and stream operations will receive.
+    *
+    * @param suppressed Configuration object determining what, if any, updates to suppress
+    * @return A new KTable with the desired suppression characteristics.
```
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala]:
```suggestion
    * @return A new [[KTable]] with the desired suppression characteristics.
```

**Review by vvcephei — CHANGES_REQUESTED** [review]
Hi @ouertani ,

Thanks for picking this up!

I commented on a couple of scaladoc syntax things.

The tests failed because the method header is missing some spaces. You can look at the console output of the build to see what it should be, or you can just run `./gradlew spottlessApply`, and it'll reformat it for you.

Experience has taught us that we really need unit tests for the scala API as well, even thought your implementation is "obviously correct". For example, in the past, we've seen problems with the type bounds.

I think a simple test that verifies:
* `untilWindowCloses` with a `KTable[Windowed[_],_]`
* `untilTimeLimit` with a `KTable[Windowed[_],_]`
* `untilTimeLimit` with a `KTable[(not Windowed) ,_]`

And, unfortunately, we can't test code that doesn't compile, but please confirm that the following fails the type check:
* `untilWindowCloses` with a `KTable[(not Windowed),_]`

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala [hunk 3]
```diff
@@ -159,6 +159,16 @@ class KTable[K, V](val inner: KTableJ[K, V]) {
...
+    */
+  def suppress(suppressed: Suppressed[_ >: K]):KTable[K,V] = inner.suppress(suppressed)
```
Comment (ryannedolan) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala]:
```suggestion
  def suppress(suppressed: Suppressed[_ >: K]): KTable[K, V] = inner.suppress(suppressed)
```

**Review by ryannedolan — CHANGES_REQUESTED** [review]

**Comment by mjsax** [comment]
Just a side remark. If this should be part of `2.2` release, we must merge it by feature freeze deadline that is Jan 31.

**Comment by ouertani** [comment]
@guozhangwang @ryannedolan updates done

**Comment by mjsax** [comment]
@ouertani This PR cannot be review atm -- it needs to be rebased to `trunk` first. Thanks.

**Comment by mjsax** [comment]
\cc @vvcephei for review (after PR is rebased).

**Comment by casey-green** [comment]
I tried using this today, so +1000 on getting this change in 😄

**Comment by guozhangwang** [comment]
@ouertani are you still working on this PR? If yes could you rebase it and then ping us for reviews?

**Comment by casey-green** [comment]
@guozhangwang @vvcephei I created a different PR for this since this one seems stuck https://github.com/apache/kafka/pull/6314

Unit tests are included in that one!

**Comment by mjsax** [comment]
What is the status of this PR?

**Comment by casey-green** [comment]
@mjsax this can be closed, I implemented this functionality in #6314

**Comment by kkonstantine** [comment]
Closing in favor of https://github.com/apache/kafka/pull/6314

**Pull request closed without merging**